crates/e2e-test-utils/src/wallet.rs

```diff
@@ -36,17 +66,38 @@ impl Wallet {
         for idx in 0..self.amount {
             let builder =
                 builder.clone().derivation_path(&format!("{derivation_path}{idx}")).unwrap();
-            let wallet = builder.build().unwrap().with_chain_id(Some(self.chain_id));
+            let inner = builder.build().unwrap().with_chain_id(Some(self.chain_id));
+            let wallet = Wallet::new(inner, self.chain_id);
             wallets.push(wallet)
         }
         wallets
     }
 }
+/// Helper struct that wraps interaction to a local wallet and a transaction generator
+pub struct Wallet {
+    inner: LocalWallet,
+    pub tx_gen: TransactionTestContext,
+}
```

Comment:
wallet needs to be nonce aware, so we can manipulate it. there are scenarios where we might want to simulate a block reorg, and so, we need to decrement it.